{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook contains code for doing GWAS regressions using the PyStan software."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import json\n",
    "import pystan\n",
    "from scipy import stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "growth_df = pd.read_csv(\"../Data/Growth_data/Cross_tolerance/Processed_growth_rates.tsv\", sep=\"\\t\", index_col=0)\n",
    "\n",
    "with open(\"../Data/Mutation_data/Strain_to_genes.json\") as infile:\n",
    "    genotypes = json.load(infile)\n",
    "gene_to_strains = {}\n",
    "for s, genes in genotypes.items():\n",
    "    for gene in genes:\n",
    "        gene_to_strains.setdefault(gene, []).append(s)\n",
    "gene_list = sorted(gene_to_strains)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['acrB',\n",
       " 'alaA',\n",
       " 'alaE',\n",
       " 'allD',\n",
       " 'ampC',\n",
       " 'amyA',\n",
       " 'argG',\n",
       " 'argS',\n",
       " 'argV',\n",
       " 'argY',\n",
       " 'arpA',\n",
       " 'atpI',\n",
       " 'barA',\n",
       " 'bglF',\n",
       " 'bglG',\n",
       " 'bioB',\n",
       " 'borD',\n",
       " 'ccmA',\n",
       " 'cheR',\n",
       " 'clpP',\n",
       " 'clpX',\n",
       " 'cmtB',\n",
       " 'crl',\n",
       " 'csiD',\n",
       " 'cspC',\n",
       " 'cspE',\n",
       " 'cydA',\n",
       " 'cynR',\n",
       " 'dacA',\n",
       " 'dcd',\n",
       " 'dcm',\n",
       " 'dcyD',\n",
       " 'dosP',\n",
       " 'dsrA',\n",
       " 'dsrB',\n",
       " 'dusA',\n",
       " 'dusB',\n",
       " 'edd',\n",
       " 'elfA',\n",
       " 'elfD',\n",
       " 'emrA',\n",
       " 'emrB',\n",
       " 'emrY',\n",
       " 'epmB',\n",
       " 'essD',\n",
       " 'eyeA',\n",
       " 'fabR',\n",
       " 'fadB',\n",
       " 'fbaB',\n",
       " 'fimC',\n",
       " 'fimD',\n",
       " 'fliA',\n",
       " 'fliC',\n",
       " 'fliD',\n",
       " 'fliE',\n",
       " 'fliF',\n",
       " 'fliG',\n",
       " 'fliH',\n",
       " 'fliI',\n",
       " 'fliJ',\n",
       " 'fliK',\n",
       " 'fliL',\n",
       " 'fliM',\n",
       " 'fliN',\n",
       " 'fliO',\n",
       " 'fliP',\n",
       " 'fliQ',\n",
       " 'fliR',\n",
       " 'fliS',\n",
       " 'fliT',\n",
       " 'fliY',\n",
       " 'fliZ',\n",
       " 'flu',\n",
       " 'focA',\n",
       " 'gabP',\n",
       " 'gatY',\n",
       " 'glmS',\n",
       " 'glmU',\n",
       " 'glnE',\n",
       " 'glnG',\n",
       " 'glsB',\n",
       " 'gltP',\n",
       " 'gltV',\n",
       " 'gltW',\n",
       " 'glxK',\n",
       " 'glyQ',\n",
       " 'glyV',\n",
       " 'glyX',\n",
       " 'glyY',\n",
       " 'greA',\n",
       " 'gtrS',\n",
       " 'gudD',\n",
       " 'hcaD',\n",
       " 'hfq',\n",
       " 'hldE',\n",
       " 'hns',\n",
       " 'hofM',\n",
       " 'ibsE',\n",
       " 'icd',\n",
       " 'icdC',\n",
       " 'idnR',\n",
       " 'ilvH',\n",
       " 'ilvN',\n",
       " 'infA',\n",
       " 'infB',\n",
       " 'insA',\n",
       " 'insH1',\n",
       " 'insI1',\n",
       " 'insN',\n",
       " 'intE',\n",
       " 'intG',\n",
       " 'iscR',\n",
       " 'kgtP',\n",
       " 'kup',\n",
       " 'lacY',\n",
       " 'lacZ',\n",
       " 'leuA',\n",
       " 'leuL',\n",
       " 'lexA',\n",
       " 'lhr',\n",
       " 'ligA',\n",
       " 'lit',\n",
       " 'lon',\n",
       " 'lrhA',\n",
       " 'lsrF',\n",
       " 'lsrG',\n",
       " 'lsrK',\n",
       " 'lysQ',\n",
       " 'lysU',\n",
       " 'lysW',\n",
       " 'lysY',\n",
       " 'lysZ',\n",
       " 'malQ',\n",
       " 'manY',\n",
       " 'marC',\n",
       " 'mdtJ',\n",
       " 'mdtK',\n",
       " 'metJ',\n",
       " 'metL',\n",
       " 'mgrB',\n",
       " 'mltD',\n",
       " 'mngB',\n",
       " 'mpl',\n",
       " 'mppA',\n",
       " 'mprA',\n",
       " 'mrdA',\n",
       " 'mrdB',\n",
       " 'mreB',\n",
       " 'mreC',\n",
       " 'murA',\n",
       " 'murC',\n",
       " 'murG',\n",
       " 'nadR',\n",
       " 'nagA',\n",
       " 'nagC',\n",
       " 'nanK',\n",
       " 'narP',\n",
       " 'nmpC',\n",
       " 'nrdE',\n",
       " 'nrdF',\n",
       " 'nrdH',\n",
       " 'nrdI',\n",
       " 'nusA',\n",
       " 'nusG',\n",
       " 'ompC',\n",
       " 'ompN',\n",
       " 'ompR',\n",
       " 'opgH',\n",
       " 'otsB',\n",
       " 'pagP',\n",
       " 'pdxJ',\n",
       " 'pepA',\n",
       " 'perR',\n",
       " 'pgrR',\n",
       " 'pheU',\n",
       " 'pnp',\n",
       " 'polB',\n",
       " 'prfA',\n",
       " 'prlF',\n",
       " 'prmA',\n",
       " 'proA',\n",
       " 'proB',\n",
       " 'proQ',\n",
       " 'proV',\n",
       " 'proW',\n",
       " 'proX',\n",
       " 'prpE',\n",
       " 'pstA',\n",
       " 'pstB',\n",
       " 'pstS',\n",
       " 'ptrA',\n",
       " 'ptsP',\n",
       " 'purL',\n",
       " 'purP',\n",
       " 'purT',\n",
       " 'pykF',\n",
       " 'pyrE',\n",
       " 'queE',\n",
       " 'rcsA',\n",
       " 'rcsD',\n",
       " 'recE',\n",
       " 'relA',\n",
       " 'rhaT',\n",
       " 'rhmD',\n",
       " 'rho',\n",
       " 'rhsA',\n",
       " 'rlmD',\n",
       " 'rlmE',\n",
       " 'rlmH',\n",
       " 'rnb',\n",
       " 'rne',\n",
       " 'rnt',\n",
       " 'rob',\n",
       " 'rph',\n",
       " 'rpoA',\n",
       " 'rpoB',\n",
       " 'rpoC',\n",
       " 'rpoD',\n",
       " 'rpoS',\n",
       " 'rpsA',\n",
       " 'rpsC',\n",
       " 'rpsD',\n",
       " 'rpsG',\n",
       " 'rpsI',\n",
       " 'rpsL',\n",
       " 'rraA',\n",
       " 'rrfE',\n",
       " 'rrfF',\n",
       " 'rrfG',\n",
       " 'rrlA',\n",
       " 'rrlE',\n",
       " 'rrlG',\n",
       " 'rrlH',\n",
       " 'rrrD',\n",
       " 'rrsA',\n",
       " 'rrsE',\n",
       " 'rrsG',\n",
       " 'rseX',\n",
       " 'rsfS',\n",
       " 'rsmG',\n",
       " 'rspA',\n",
       " 'rzoD',\n",
       " 'rzpD',\n",
       " 'sad',\n",
       " 'sapA',\n",
       " 'sapB',\n",
       " 'sapC',\n",
       " 'sapD',\n",
       " 'sapF',\n",
       " 'sdaC',\n",
       " 'sdiA',\n",
       " 'serX',\n",
       " 'sfmH',\n",
       " 'speA',\n",
       " 'spoT',\n",
       " 'srmB',\n",
       " 'sspA',\n",
       " 'ssuA',\n",
       " 'stfE',\n",
       " 'stpA',\n",
       " 'tam',\n",
       " 'tdcR',\n",
       " 'tdk',\n",
       " 'thrA',\n",
       " 'thrV',\n",
       " 'thrW',\n",
       " 'tolA',\n",
       " 'tolC',\n",
       " 'tomB',\n",
       " 'tqsA',\n",
       " 'treR',\n",
       " 'trkH',\n",
       " 'tufA',\n",
       " 'tyrB',\n",
       " 'ubiE',\n",
       " 'umuD',\n",
       " 'uvrB',\n",
       " 'uxaB',\n",
       " 'valT',\n",
       " 'valZ',\n",
       " 'vsr',\n",
       " 'waaS',\n",
       " 'wbbK',\n",
       " 'wzzE',\n",
       " 'xapR',\n",
       " 'yafW',\n",
       " 'yafX',\n",
       " 'yafY',\n",
       " 'yafZ',\n",
       " 'yagL',\n",
       " 'yagU',\n",
       " 'yaiP',\n",
       " 'ybbW',\n",
       " 'ybcK',\n",
       " 'ybcW',\n",
       " 'ybeX',\n",
       " 'ybgT',\n",
       " 'ybjL',\n",
       " 'ycaN',\n",
       " 'ycdU',\n",
       " 'ycfQ',\n",
       " 'ycgB',\n",
       " 'yciA',\n",
       " 'ycjG',\n",
       " 'ydbA',\n",
       " 'ydcD',\n",
       " 'ydcI',\n",
       " 'ydeA',\n",
       " 'ydfI',\n",
       " 'ydfJ',\n",
       " 'ydhZ',\n",
       " 'ydiJ',\n",
       " 'ydjH',\n",
       " 'yeaR',\n",
       " 'yebO',\n",
       " 'yecC',\n",
       " 'yecF',\n",
       " 'yecS',\n",
       " 'yecT',\n",
       " 'yedA',\n",
       " 'yedD',\n",
       " 'yedE',\n",
       " 'yedF',\n",
       " 'yedI',\n",
       " 'yedJ',\n",
       " 'yedK',\n",
       " 'yedL',\n",
       " 'yedN',\n",
       " 'yedP',\n",
       " 'yedQ',\n",
       " 'yedR',\n",
       " 'yedS',\n",
       " 'yedV',\n",
       " 'yehD',\n",
       " 'yehE',\n",
       " 'yfcZ',\n",
       " 'yffQ',\n",
       " 'yfgF',\n",
       " 'yfgO',\n",
       " 'yfiN',\n",
       " 'yfjL',\n",
       " 'yfjM',\n",
       " 'yfjW',\n",
       " 'ygaC',\n",
       " 'ygaH',\n",
       " 'ygaM',\n",
       " 'ygaQ',\n",
       " 'ygaY',\n",
       " 'ygaZ',\n",
       " 'ygbT',\n",
       " 'ygcE',\n",
       " 'ygeG',\n",
       " 'ygjP',\n",
       " 'yhbY',\n",
       " 'yhgE',\n",
       " 'yhiL',\n",
       " 'yhjA',\n",
       " 'yhjK',\n",
       " 'yiaT',\n",
       " 'yiaU',\n",
       " 'yicC',\n",
       " 'yieK',\n",
       " 'yihQ',\n",
       " 'yihR',\n",
       " 'yijD',\n",
       " 'yjcE',\n",
       " 'yjcF',\n",
       " 'yjhG',\n",
       " 'yjiP',\n",
       " 'yjjQ',\n",
       " 'ykfA',\n",
       " 'ykfB',\n",
       " 'ykfC',\n",
       " 'ykfF',\n",
       " 'ykfG',\n",
       " 'ykfH',\n",
       " 'ykfI',\n",
       " 'ykfK',\n",
       " 'ykfL',\n",
       " 'ykfN',\n",
       " 'ykgI',\n",
       " 'ylcI',\n",
       " 'ymfE',\n",
       " 'ynaI',\n",
       " 'yneE',\n",
       " 'yneF',\n",
       " 'yneG',\n",
       " 'yneJ',\n",
       " 'yneK',\n",
       " 'yneO',\n",
       " 'yobF',\n",
       " 'yobH',\n",
       " 'yodC',\n",
       " 'yodD',\n",
       " 'yphC',\n",
       " 'yphF',\n",
       " 'ypjA',\n",
       " 'ypjC',\n",
       " 'ytfR']"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "gwas_df = growth_df.copy()\n",
    "gwas_df = gwas_df.groupby([\"strain\", \"grown_compound\"]).mean().reset_index()\n",
    "gwas_df = gwas_df[pd.notnull(gwas_df[\"rel_growth_rate\"])]\n",
    "gwas_genes = []\n",
    "for gene in gene_list:\n",
    "    if len(gene_to_strains[gene]) > 1:\n",
    "        gwas_genes.append(gene)\n",
    "        gwas_df[gene] = gwas_df[\"strain\"].map(lambda x: 1 if x in gene_to_strains[gene] else 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Define the regression model\n",
    "\n",
    "stan_model = \"\"\"\n",
    "data {\n",
    "    int N; // Number of observations\n",
    "    int M; // Number of genes\n",
    "    vector[N] y; // Growth rates\n",
    "    matrix[N, M] X; // Mutation data\n",
    "}\n",
    "\n",
    "parameters {\n",
    "    vector[M] beta; // coefficients\n",
    "    real intercept; // intercept of the fit\n",
    "    real log_sigma; // Measuring STD\n",
    "    real log_tau; // Hyperparameter: STD of prior on betas\n",
    "}\n",
    "\n",
    "model {\n",
    "    // increment_log_prob(normal_log(beta, 0, exp(log_tau)) + double_exponential_log(beta, 0, exp(log_tau)));\n",
    "    // beta ~ double_exponential(0, exp(log_tau));\n",
    "    beta ~ normal(0, exp(log_tau));\n",
    "    y ~ normal(intercept + X*beta, exp(log_sigma));\n",
    "}\n",
    "\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:pystan:COMPILING THE C++ CODE FOR MODEL anon_model_ad6b17b3e0394de93c979fff36836b1b NOW.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HMDA\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "INFO:pystan:COMPILING THE C++ CODE FOR MODEL anon_model_ad6b17b3e0394de93c979fff36836b1b NOW.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "putrescine\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "INFO:pystan:COMPILING THE C++ CODE FOR MODEL anon_model_ad6b17b3e0394de93c979fff36836b1b NOW.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1,2-propanediol\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "INFO:pystan:COMPILING THE C++ CODE FOR MODEL anon_model_ad6b17b3e0394de93c979fff36836b1b NOW.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2,3-butanediol\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "INFO:pystan:COMPILING THE C++ CODE FOR MODEL anon_model_ad6b17b3e0394de93c979fff36836b1b NOW.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "glutarate\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "INFO:pystan:COMPILING THE C++ CODE FOR MODEL anon_model_ad6b17b3e0394de93c979fff36836b1b NOW.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "adipate\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "INFO:pystan:COMPILING THE C++ CODE FOR MODEL anon_model_ad6b17b3e0394de93c979fff36836b1b NOW.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hexanoate\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "INFO:pystan:COMPILING THE C++ CODE FOR MODEL anon_model_ad6b17b3e0394de93c979fff36836b1b NOW.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "octanoate\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "INFO:pystan:COMPILING THE C++ CODE FOR MODEL anon_model_ad6b17b3e0394de93c979fff36836b1b NOW.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "isobutyrate\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "INFO:pystan:COMPILING THE C++ CODE FOR MODEL anon_model_ad6b17b3e0394de93c979fff36836b1b NOW.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coumarate\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "INFO:pystan:COMPILING THE C++ CODE FOR MODEL anon_model_ad6b17b3e0394de93c979fff36836b1b NOW.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "butanol\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n",
      "/Users/krisj/anaconda/envs/tolerance/lib/python3.5/multiprocessing/reduction.py:50: UserWarning: Pickling fit objects is an experimental feature!\n",
      "The relevant StanModel instance must be pickled along with this fit object.\n",
      "When unpickling the StanModel must be unpickled first.\n",
      "  cls(buf, protocol).dump(obj)\n"
     ]
    }
   ],
   "source": [
    "compound_list = [\n",
    "    \"HMDA\", \"putrescine\", \"1,2-propanediol\", \"2,3-butanediol\",\n",
    "    \"glutarate\", \"adipate\", \"hexanoate\", \"octanoate\", \"isobutyrate\", \"coumarate\", \"butanol\"]\n",
    "\n",
    "chains = 4\n",
    "\n",
    "stan_data_list = []\n",
    "\n",
    "fits = {}\n",
    "for comp in compound_list:\n",
    "    print(comp)\n",
    "    compound_df = gwas_df[gwas_df[\"grown_compound\"] == comp]\n",
    "\n",
    "    y = compound_df[\"rel_growth_rate\"].values\n",
    "    X = compound_df[gwas_genes]\n",
    "    X = X - X.mean()\n",
    "    X = X.values\n",
    "\n",
    "    stan_data = {\n",
    "        \"N\": len(y),\n",
    "        \"M\": X.shape[1],\n",
    "        \"y\": y,\n",
    "        \"X\": X\n",
    "    }\n",
    "    stan_data_list.append(stan_data)\n",
    "    fit = pystan.stan(model_code=stan_model, data=stan_data, chains=chains, n_jobs=chains, iter=500)\n",
    "    fits[comp] = fit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def calc_significance(ar, b=0):\n",
    "    perc = stats.percentileofscore(ar, b)\n",
    "    perc = min(perc, 100-perc)\n",
    "    perc = perc*2\n",
    "    perc = perc / 100\n",
    "    if perc == 0:\n",
    "        perc = 1/len(ar)\n",
    "    return perc\n",
    "\n",
    "p_values = {}\n",
    "coef_means = {}\n",
    "for comp, fit in fits.items():\n",
    "    samples = fit.extract()\n",
    "    pvals = dict(pd.Series((calc_significance(samples[\"beta\"][:, i]) for i in range(len(gwas_genes))), index=gwas_genes))\n",
    "    p_values[comp] = pvals\n",
    "    means = dict(pd.Series(samples[\"beta\"].mean(0), index=gwas_genes))\n",
    "    coef_means[comp] = means\n",
    "        \n",
    "p_val_df = pd.DataFrame(p_values)[compound_list]\n",
    "coef_mean_df = pd.DataFrame(coef_means)[compound_list]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "info = \"\"\"\n",
    "Gene coefficients estimated using Bayesian regression, with a normal prior (centered on 0) on coefficients.\n",
    "Posterior 'pvalues' are reported, calculated as 1 - largest credible interval that contains 0.\n",
    "\"\"\"\n",
    "writer = pd.ExcelWriter(\"../Data/GWAS/Relative_diff_Bayesian_ridge_regression_coefs.xlsx\")\n",
    "\n",
    "p_val_df.to_excel(writer, sheet_name=\"posterior p_vals\")\n",
    "coef_mean_df.to_excel(writer, sheet_name=\"coef means\")\n",
    "\n",
    "pd.DataFrame({\"Info\": [info]}).to_excel(writer, sheet_name=\"info\")\n",
    "writer.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "writer = pd.ExcelWriter(\"../Data/GWAS/Coefficients_and_significance.xlsx\")\n",
    "\n",
    "comp_conversion = {\n",
    "    \"1,2-propanediol\": \"12PD\", \"2,3-butanediol\": \"23BD\", \"putrescine\": \"PUTR\", \"HMDA\": \"HMDA\", \"glutarate\": \"GLUT\",\n",
    "    \"adipate\": \"ADIP\", \"hexanoate\": \"HEXA\", \"octanoate\": \"OCTA\", \"coumarate\": \"COUM\", \"isobutyrate\": \"IBUA\",\n",
    "    \"butanol\": \"BUT\"\n",
    "}\n",
    "\n",
    "for comp in fits:\n",
    "    comp_df = pd.DataFrame({\"mean\": coef_mean_df[comp], \"significance\": p_val_df[comp]})\n",
    "    #comp_df[\"found\"] = comp_df.index.map(keio_genes[comp_conversion[comp]].__contains__)\n",
    "    comp_df[\"count\"] = comp_df.index.map(lambda x: len(gene_to_strains[x]))\n",
    "    #comp_df[\"mutations\"] = comp_df.index.map(lambda x: \", \".join(gene_name_to_mutations[x]))\n",
    "    comp_df = comp_df.sort_values(\"mean\", ascending=False)\n",
    "    comp_df.to_excel(writer, comp)\n",
    "\n",
    "writer.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [tolerance]",
   "language": "python",
   "name": "Python [tolerance]"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
